### PR TITLE
Remove queries from the driver interface

### PR DESCRIPTION
**Pull request by bjorn3**
All uses of driver queries in the public api of rustc_driver have been removed in https://github.com/rust-lang/rust/pull/134130 already. This removes driver queries from rustc_interface and does a couple of cleanups around TyCtxt construction and entering enabled by this removal.

Finishes the removal of driver queries started with https://github.com/rust-lang/rust/pull/126834.

### CODE REVIEW

File: compiler/rustc_interface/src/passes.rs
Comment (bjorn3) [inline compiler/rustc_interface/src/passes.rs:778]: This is ugly, but the only way I could figure out to get all invariant lifetimes correctly inferred by rustc.

File: compiler/rustc_interface/src/passes.rs
Comment (bjorn3) [inline compiler/rustc_interface/src/passes.rs:815]: This fixme is a pre-existing issue.

File: compiler/rustc_driver_impl/Cargo.toml
Comment (klensy) [inline compiler/rustc_driver_impl/Cargo.toml:22]: Is this used? can't find any new `rustc_hir` around.

Comment (bjorn3) [reply compiler/rustc_driver_impl/Cargo.toml]: Doesn't look like it, removed.

**Comment by jieyouxu** [comment]
r? jieyouxu

**Review by jieyouxu — APPROVED** [review]
Thanks, this looks very reasonable to me as a first-pass review. This is tricky enough that I want a second pair of eyes on it.

**Comment by oli-obk** [comment]
Yea I already reviewed all of it but the lifetime stuff. But since at best I'd have some golfing for it, r=me,jieyouzu I can golf once it's landed

**Comment by jieyouxu** [comment]
In that case, I say we try to land this first, the lifetimes in passes are a bit annoying but I don't think is worth blocking, as oli said.

@bors r=oli-obk,jieyouxu rollup=never (tricky driver changes)

**Comment by jieyouxu** [comment]
@bjorn3 if you feel like it, can you double-check the `rustc_driver`-related docs in rustc-dev-guide?

**Comment by bjorn3** [comment]
I'm going to update the rustc dev guide.

**Comment by jieyouxu** [comment]
> The job **`x86_64-msvc`** failed! Check out the build log: [(web)](https://github.com/rust-lang-ci/rust/actions/runs/12345462617/job/34449612865) [(plain)](https://github.com/rust-lang-ci/rust/commit/38de828a35adeab40e2ee8e1f0c2e1d3b437b776/checks/34449612865/logs)
> Click to see the possible cause of the failure (guessed by this bot)

cc #132111 and #133432.

**Comment by Kobzol** [comment]
This seems to have regressed `doc` builds (up to ~2.5% on primary benchmark doc builds). Was that expected?

File: src/librustdoc/lib.rs
Comment (bjorn3) [inline src/librustdoc/lib.rs:862]: Looks like this accidentally got moved out of the `create_and_enter_global_ctxt()`.

**Comment by bjorn3** [comment]
No regressions of any kind were expected.

**Comment by Kobzol** [comment]
Ok. Given that the regressions are tiny and only in doc builds, I think that it is fine.

@rustbot label: +perf-regression-triaged

File: src/librustdoc/lib.rs
Comment (lqd) [reply src/librustdoc/lib.rs:862]: Would this be related to the rustdoc regression?

Comment (bjorn3) [reply src/librustdoc/lib.rs:862]: I don't think so. In any case opened https://github.com/rust-lang/rust/pull/135157 to revert this code movement.

**Comment by marxin** [comment]
@bjorn3 
I noticed an orphan link in `rustc-dev-guide`:

```
error: Server returned 404 Not Found for https://doc.rust-lang.org/nightly/nightly-rustc/rustc_interface/queries/struct.Query.html
   ┌─ appendix/code-index.md:21:194
   │
21 │ `Query` | struct | Represents the result of query to the `Compiler` interface and allows stealing, borrowing, and returning the results of compiler passes. | [The Rustc Driver and Interface] | [compiler/rustc_interface/src/queries.rs](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_interface/queries/struct.Query.html)
   │                                                                                                                                                                                                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ Server returned 404 Not Found for https://doc.rust-lang.org/nightly/nightly-rustc/rustc_interface/queries/struct.Query.html
```

**Comment by bjorn3** [comment]
That was fixed in https://github.com/rust-lang/rustc-dev-guide/commit/7a4c3d3c30d5dcff62722bfb9edf963e4d8c0c77.

**Pull request merged**
